src/evm-block-extractor/src/database/postgres_db_client.rs

```diff
@@ -237,6 +240,94 @@ impl DatabaseClient for PostgresDbClient {
             .map_err(|e| anyhow::anyhow!("Error getting transaction {}: {:?}", hex_tx_hash, e))
             .and_then(|row| from_row_value(&row, 0))
     }
+
+    async fn discard_blocks_starting_with(
+        &self,
+        start_from: u64,
+        reason: &str,
+    ) -> anyhow::Result<()> {
+        log::warn!("Discarding blocks starting with {start_from}");
+
+        let mut tx = self.pool.begin().await?;
+        sqlx::query(
+            "
+            WITH deleted_txs AS (
+            	DELETE FROM evm_transaction
+            	WHERE block_number >= $1
+            	RETURNING *
+            ) 
+            INSERT INTO discarded_evm_transaction 
+            SELECT * FROM deleted_txs
+            ",
+        )
+        .bind(start_from as i64)
+        .execute(&mut *tx)
+        .await
+        .map_err(|e| anyhow::anyhow!("Failed to discard tail: {e}"))?;
+
+        sqlx::query(
+            "
+            WITH deleted_blocks AS (
+            	DELETE FROM evm_block
+            	WHERE id >= $1
+            	RETURNING *
+            ) 
+            INSERT INTO discarded_evm_block (id, data, reason, discarded_at)
+            SELECT id, data, '$2', now() FROM deleted_blocks;
+            ",
+        )
+        .bind(start_from as i64)
+        .bind(reason)
+        .execute(&mut *tx)
+        .await
+        .map_err(|e| anyhow::anyhow!("Failed to discard tail: {e}"))?;
+
+        sqlx::query("DELETE FROM certified_evm_block WHERE id >= $1")
+            .bind(start_from as i64)
+            .execute(&mut *tx)
+            .await
+            .map_err(|e| anyhow::anyhow!("Failed to discard tail: {e}"))?;
+
+        tx.commit().await?;
+
+        Ok(())
+    }
+
+    async fn get_discarded_block_by_number(&self, number: u64) -> anyhow::Result<DiscardedBlock> {
```

Comment:
I think the ID is not meaningfull in this scope, it could be more useful to get the blocks by hash or by block number.

Reply:
Currently, the ID is `block_number`. But you right, it is possible that different blocks with the same block numbers can be discarded. So, I'll better use block hashes. Block number will be an additional index.

Reply:
Using block hash as ID now